## LIBRARIES
  suppressPackageStartupMessages({
    library(shiny)
    library(plotly)
    library(iheatmapr)
    library(shinyWidgets)
    library(shinyBS)
    library(colourpicker)
  })

## ANNOTATIONS AND DATA REQUIRED FOR UI
  #MAIN_PATH = '/Users/nicco/Documents/GitHub/snpXplorer/bin/snpxplorer_v4/'
  MAIN_PATH = '/root/snpXplorer/'
  gtex_tissues = data.table::fread(paste0(MAIN_PATH, 'data/databases/tissues_gtex.txt'), h=F, sep = "\t")

## USER INTERFACE
  navbarPage("snpXplorer", inverse = T,
    tabPanel("Exploration",             # Exploration panel
      fluidRow(column(12, img(src='dna.png', align="center", width="100%"))),             # place DNA image on top of everything -- this is in a single row 
      fluidRow(column(12, img(src='tutorial.png', align="center", width="100%"))),        # then place in another row with the tutorial information
      hr(),
      fluidRow(           # INPUT SELECTION -- BROWSING OPTIONS -- REFERENCE GENOME
        column(3,                                                                       # first column will be about the available input data to visualize (datasets or own dataset)
          wellPanel(
            div(tags$h2("Available GWAS"), align = "center"),                                               # name of the dropdown menu
            div(dropdown(                                                                     # dropdown menu here
              style = "unite", circle = F, size = "lg", icon = icon("database"), status = "default", width = "500px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
              tooltip = tooltipOptions(title = "Click to see available options!"),
              h6("Please check up to 5 boxes at the same time."),                         # additional text here and next line
              h6("By default (i.e. without selecting any input dataset, snpXplorer will load IGAP study (chr16)."),
              prettyCheckboxGroup(inputId = "neurol_gwas", label = "Neurological GWAS:", choices = c("Alzheimer Million" = "Alzheimer_million", "Alzheimer GR@ACE" = "GR@ACE", "Alzheimer IGAP" = "IGAP", "Alzheimer Proxy" = "proxy_AD", "Autism", "Depression", "Ventricular volume" = "LVV"), inline = TRUE, status = "default", fill = TRUE),
              prettyCheckboxGroup(inputId = "cardio_gwas", label = "Cardiovascular GWAS:", choices = c("CAD", "SBP", "BMI", "Diabetes"), inline = TRUE, status = "danger", fill = TRUE),
              prettyCheckboxGroup(inputId = "immune_gwas", label = "Immunological GWAS:", choices = c("COVID", "Lupus", "Inflammation", "Asthma"), inline = TRUE, status = "success", fill = TRUE),
              prettyCheckboxGroup(inputId = "cancer_gwas", label = "Oncological GWAS:", choices = c("Breast cancer" = "Breast_cancer", "Myeloproliferative", "Prostate", "Lung cancer" = "Lung", "Leukemia"), inline = TRUE, status = "warning", fill = TRUE),
              prettyCheckboxGroup(inputId = "physio_gwas", label = "Physiological GWAS:", choices = c("Multivariate Longevity" = "Multivariate_Longevity", "Parental Longevity" = "UKBaging", "Height", "Education", "Bone density" = "Bone_density", "Vitamin D" = "Vitamin_D"), inline = TRUE, status = "info", fill = TRUE),
              bsPopover("neurol_gwas", title='Neurological GWAS', content='Alzheimer Million: GWAS of Alzheimer including 1 million individuals -- Alzheimer GR@ACE: Genomic Research @ Fundacio ACE (GWAS of Alzheimers disease) -- Alzheimer IGAP: International Genomics of Alzheimers disease (AD, case-control GWAS) -- Alzheimer Proxy: family history of AD', placement="right",trigger="hover", options = list(container = "body")),
              bsPopover("cardio_gwas", title='Cardiovascular GWAS', content='CAD: Coronary Artery Disease -- SBP: Systolic Blood Pressure -- BMI: Body-mass index', placement="right",trigger="hover", options = list(container = "body")),
              bsPopover("immune_gwas", title='Immunological GWAS', content='COVID: severe CODIV-19 reactions -- Lupus: Lupus Erithematosous', placement="right",trigger="hover", options = list(container = "body")),
              bsPopover("physio_gwas", title='Physiological GWAS', content='Multivariate Longevity: Multivariate analysis of Lifespan, Longevity and Healthspan', placement="right",trigger="hover", options = list(container = "body")),
            ), align = "center"),
          ),
        ),
        column(3,                                                                       # second column is for the input file from the user
          wellPanel(
            div(tags$h2("Add your GWAS"), align = "center"),
            div(dropdown(
              style = "unite", size = "lg", icon = icon("upload"), status = "danger", width = "500px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
              tooltip = tooltipOptions(title = "Click to see available options!"),
              h3("Visualize your own GWAS data"),
              prettyCheckboxGroup(inputId = "owned_gwas", label = "Visualize your own GWAS:", choices = c("Show your data"), inline = TRUE, status = "primary", fill = TRUE),
              fileInput(inputId = "upload_file", label = "Upload your data:", multiple = F, placeholder = "Loading file", width="300px"),
              h5('Please make sure data has at least chromosome, position and p-value columns. A header is required. See More/Help to download sample files and More/Documentation for additional information.'),
            ), align = "center"),
          ),
        ),
        column(3,                                                                       # third column is for the reference genome
          wellPanel(
            div(tags$h2("Reference Genome"), align = "center"),                                              # name of the dropdown menu
            div(dropdown(
              style = "unite", size = "lg", icon = icon("asterisk"), status = "warning", width = "500px", animate = animateOptions(enter = animations$fading_entrances$fadeInRightBig, exit = animations$fading_exits$fadeOutRightBig),
              tooltip = tooltipOptions(title = "Click to see available options!"),
              h3("Choose your reference genome"),
              prettyRadioButtons(inputId = "reference_gen", label = "Available versions:", choices = c("GRCh37 (hg19)", "GRCh38 (hg38)"), inline = TRUE, status = "info", fill = TRUE),
              h5('By default, GRCh37 (hg19) is used. snpXplorer uses liftOver tool to change genomic coordinates between reference genome versions.'),
            ), align = "center"),
          ),
        ),
        column(3,                                                                       # fourth column is for the browsing options
          wellPanel(
            div(tags$h2("Browsing options"), align = "center"),                                              # name of the dropdown menu
            div(dropdown(
              style = "unite", size = "lg", icon = icon("search"), status = "success", width = "300px", animate = animateOptions(enter = animations$fading_entrances$fadeInRightBig, exit = animations$fading_exits$fadeOutRightBig),
              tooltip = tooltipOptions(title = "Click to see available options!"),
              h3("Choose how to browse the genome"),
              prettyRadioButtons(inputId = "browsing_options", label = "Available options:", choices = c("Locus, Gene, RsID", "Manual Scroll"), icon = icon("check"), bigger = TRUE, status = "warning", animation = "jelly"),
              conditionalPanel(condition = "input.browsing_options == 'Locus, Gene, RsID'",            # conditional panel for the 'Locus, Gene, RsID mode'
                  textInput(inputId = "target", label = "Insert your target region", value = "Type locus, rsID or gene name", width="400px"),
                  h5("Locus: genomic location of interest. The format is chromosome:position (e.g 1:12345678)"),
                  h5("RsID: variant of interest. The format is rsXXXX (e.g rs7412)"),
                  h5("Gene name: gene of interest. The format is gene symbol (e.g APOE)")),
              conditionalPanel(condition = "input.browsing_options == 'Manual Scroll'",                # conditional panel for the 'Manual scroll'
                  sliderInput(inputId = 'manual_chrom', label = "Chromosome", value = 11, min=1, max=22, width = "100%"),
                  sliderInput(inputId = 'manual_pos', label = "Position", 15000000, min = 1, max = 250000000, width = "100%")),
            ), align = "center"),
          ),
        ),
      ),
      fluidRow(
        column(12,
          div(submitButton(text = "Show Results!", icon("dna"), width="300px"), style = "font-size:200%", align = "center"),
        ),
      ),
      hr(),
      fluidRow(           # MAIN PLOT WINDOW AND SIDEBAR OPTIONS
        column(4,
          wellPanel(style = "background: #f2f2f2",
            column(6,
              div(tags$h3("Graphical options"), align = "center"),
              div(dropdown(
                style = "unite", size = "lg", icon = icon("gear"), status = "success", width = "500px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
                tooltip = tooltipOptions(title = "Click to see available options!"),
                h3("Customize your plot"),
                colourpicker::colourInput("col1", "Select colour", "blue"),             # color selection for the different input gwases
                conditionalPanel(condition = "(input.neurol_gwas.length + input.cardio_gwas.length + input.immune_gwas.length + input.cancer_gwas.length + input.physio_gwas.length + input.owned_gwas.length) > 1",
                  colourpicker::colourInput("col2", "Select color 2", "coral")),
                conditionalPanel(condition = "(input.neurol_gwas.length + input.cardio_gwas.length + input.immune_gwas.length + input.cancer_gwas.length + input.physio_gwas.length + input.owned_gwas.length) > 2",
                  colourpicker::colourInput("col3", "Select color 3", "black")),
                conditionalPanel(condition = "(input.neurol_gwas.length + input.cardio_gwas.length + input.immune_gwas.length + input.cancer_gwas.length + input.physio_gwas.length + input.owned_gwas.length) > 3",
                  colourpicker::colourInput("col4", "Select color 4", "green")),
                conditionalPanel(condition = "(input.neurol_gwas.length + input.cardio_gwas.length + input.immune_gwas.length + input.cancer_gwas.length + input.physio_gwas.length + input.owned_gwas.length) > 4",
                  colourpicker::colourInput("col5", "Select color 5", "yellow")),
                bsPopover("col", title='Colors', content='You can use the colour picker to change the color of the plotted studies.', placement="right",trigger="hover", options = list(container = "body")),
                sliderInput(inputId = "x_axis", label = "Window (bp)", value = 25000, min = 1000, max = 500000, width = "100%"),        # adjust the width of the x-axis
                bsPopover("x_axis", title='Window size', content='You can use the slider to adjust x-axis of the main plot.', placement="right",trigger="hover", options = list(container = "body")),
                sliderInput(inputId = "y_axis", label = "-log10 (P-value)", value = 9, min = 5, max = 100, width = "100%"),              # adjust the width of the y-axis
                bsPopover("y_axis", title='Window y-size', content='You can use the slider to adjust y-axis of the main plot.', placement="right",trigger="hover", options = list(container = "body")),
                radioGroupButtons(inputId = "plot_type", label = "Plot type:", choices = c("Scatter", "Density"), status = "primary", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
                bsPopover("ploType", title='Plot type', content='Points: each SNP is a dot -- Lines: draw p-value density lines.', placement="right",trigger="hover", options = list(container = "body")),
                radioGroupButtons(inputId = "recomb_yn", label = "Add recombination rates:", choices = c("No", "Yes"), status = "warning", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
                bsPopover("recomb_yn", title='Show recombination rates', content='Show (Yes) or hide (No) recombination rates from the main plot.', placement="right",trigger="hover", options = list(container = "body")),
                radioGroupButtons(inputId = "exons", label = "Add exons:", choices = c("No", "Yes"), status = "danger", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
                bsPopover("exons", title='Add gene exons', content='Show (Yes) or hide (No) exons in each gene. Note that the waiting time will be slightly higher.', placement="right",trigger="hover", options = list(container = "body")),
              ), align = "center"),
            ),
            column(6,
              div(tags$h3("Linkage Disequilibrium"), align = "center"),
              div(dropdown(
                style = "unite", size = "lg", icon = icon("link"), status = "royal", width = "500px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
                tooltip = tooltipOptions(title = "Click to see available options!"),
                h3("Add LD"),
                radioGroupButtons(inputId = "linkage_type", label = "LD type:", choices = c("No LD", "Input variant", "Most significant in window"), status = "danger", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),      # select LD type
                bsPopover("linkage_type", title='LD type', content='Define if LD should be shown (only available when 1 study is selected). If LD should be shown, please indicate in which populations (from 1000Genome Project) LD should be calculated. ', placement="right",trigger="hover", options = list(container = "body")),
                h5("Select population(s) to calculate LD"),                                   # panels for the choice of the population of interest
                prettyCheckboxGroup(inputId ="pop_eur", label = "European", choices = c("CEU" = "CEU", "FIN" = "FIN", "GBR" = "GBR", "IBS" = "IBS", "TSI" = "TSI", "ALL Sub-Pop" = "ALL_eur"), inline = TRUE, status = "info", fill = TRUE),
                bsPopover("pop_eur", title='European populations', content='CEU: Utah Residents with Northern and Western European Ancestry -- FIN: Finnish -- GBR: British in England and Scotland -- IBS: Iberian Population in Spain -- TSI: Toscani in Italy', placement="right",trigger="hover", options = list(container = "body")),
                prettyCheckboxGroup(inputId = "pop_afr", label = "African", choices = c("ACB" = "ACB", "ASW" = "ASW", "ESN" = "ESN", "GWD" = "GWD", "LWK" = "LWK", "MSL" = "MSL", "YRI" = "YRI", "ALL Sub-Pop" = "ALL_afr"), inline = TRUE, status = "warning", fill = TRUE),
                bsPopover("pop_afr", title='African populations', content='ACB: African Caribbeans in Barbados -- ASW: Americans of African Ancestry in SW USA -- ESN: Esan in Nigeria -- GWD: Gambian in Western Divisions in the Gambia -- LWK: Luhya in Webuye, Kenya -- MSL: Mende in Sierra Leone -- YRI: Yoruba in Ibadan, Nigeria', placement="right",trigger="hover", options = list(container = "body")),
                prettyCheckboxGroup(inputId = "pop_amr", label = "American", c("CLM" = "CLM", "MXL" = "MXL", "PEL" = "PEL", "PUR" = "PUR", "ALL Sub-Pop" = "ALL_amr"), inline = TRUE, status = "danger", fill = TRUE),
                bsPopover("pop_amr", title='American populations', content='CLM: Colombians from Medellin, Colombia -- MXL: Mexican Ancestry from Los Angeles USA -- PEL: Peruvians from Lima, Peru -- PUR: Puerto Ricans from Puerto Rico', placement="right",trigger="hover", options = list(container = "body")),
                prettyCheckboxGroup(inputId = "pop_eas", label = "East-Asian", choices = c("CDX" = "CDX", "CHB" = "CHB", "CHS" = "CHS", "JPT" = "JPT", "KHV" = "KHV", "ALL Sub-Pop" = "ALL_eas"), inline = TRUE, status = "success", fill = TRUE),
                bsPopover("pop_eas", title='East-Asian populations', content='CDX: Chinese Dai in Xishuangbanna, China -- CHB: Han Chinese in Beijing, China -- CHS: Southern Han Chinese -- JPT: Japanese in Tokyo, Japan -- KHV: Kinh in Ho Chi Minh City, Vietnam', placement="right",trigger="hover", options = list(container = "body")),
                prettyCheckboxGroup(inputId = "pop_sas", label = "South-Asian", choices = c("BEB" = "BEB", "GIH" = "GIH", "ITU" = "ITU", "PJL" = "PJL", "STU" = "STU", "ALL Sub-Pop" = "ALL_sas"), inline = TRUE, status = "primary", fill = TRUE),
                bsPopover("pop_sas", title='South-Asian populations', content='BEB: Bengali from Bangladesh -- GIH: Gujarati Indian from Houston, Texas -- ITU: Indian Telugu from the UK -- PJL: Punjabi from Lahore, Pakistan -- STU: Sri Lankan Tamil from the UK', placement="right",trigger="hover", options = list(container = "body")),
                h6("** If you don't select any population, by default European samples of the 1000Genome project will be used. Please refer to the Help section for additional information about the populations reported above."),
                div(style = "margin-top: 10px"),
                h6("** In case the top SNP is not in 1000Genome data, we take the second-most significant, then the third-most significant, etc."),
              ), align = "center"),
            ),
            column(6,
              div(tags$h3("Structural Variants"), align = "center"),
              div(dropdown(
                style = "unite", size = "lg", icon = icon("eraser"), status = "warning", width = "500px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
                tooltip = tooltipOptions(title = "Click to see available options!"),
                h3("Choose dataset"),
                radioGroupButtons(inputId = "strVar_inp", label = "Structural variants data", choices = c("Linthrost et al. 2019" = "jasper", "Chaisson et al. 2019" = "chaisson", "Audano et al 2019" = "audano", "All" = "all"), status = "primary", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
                bsPopover("strVar_inp", title='Structural variants', content='Third-generation sequencing (long-read sequencing) studies estimating the major Structural Variants (SV) alleles.', placement="right",trigger="hover", options = list(container = "body")),
              ), align = "center"),
            ),
            column(6,
              div(tags$h3("GWAS Catalog settings"), align = "center"),
              div(dropdown(
                style = "unite", size = "lg", icon = icon("table"), status = "danger", width = "500px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
                tooltip = tooltipOptions(title = "Click to see available options!"),
                h3("Choose dataset"),
                radioGroupButtons(inputId = "gwascat_type", label = "GWAS Catalog data", choices = c("SNPs", "Genes"), status = "primary", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
                bsPopover("gwascat_type", title='GWAS Catalog', content='Match results in GWAS Catalog by SNPs (default) or by Gene.', placement="right",trigger="hover", options = list(container = "body")),
              ), align = "center"),
            ),
            column(6,
              div(tags$h3("Download center"), align = "center"),
              div(dropdown(
                style = "unite", size = "lg", icon = icon("download"), status = "success", width = "300px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
                tooltip = tooltipOptions(title = "Click to see available options!"),
                div(h5("SNPs table"), align = "center"),
                div(downloadButton("download_SNPsTable", "Download SNPs"), align = "center"),
                hr(),
                div(h5("GWAS Catalog results"), align = "center"),
                div(downloadButton("download_GWASCat", "Download All Results"), align = "center"),
                hr(),
                div(h5("SVs in the region"), align = "center"),
                div(downloadButton("download_SVs", "Download All SVs"), align = "center"),
                hr(),
                div(h5("LD info"), align = "center"),
                div(downloadButton("download_LDtable", "Download LD table"), align = "center"),
                hr(),
                div(h5("Plot"), align = "center"),
                div(h6("You can get a .png image of the plot by hovering on the plot. Click on the camera icon!"), align = "center"),
                div(h6("Alternatively, here you can get a .pdf plot or the original .html page (preserves hovering information)."), align = "center"),
                radioGroupButtons(inputId = "plot_format", label = "", choices = c("PDF", "HTML"), status = "warning", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
                div(downloadButton("download_plot", "Download Plot"), align = "center"),
                div(h6('Note that it will take about 10 seconds to produce your plot!'), align = "center"),
              ), align = "center"),
            ),
            column(6,
              div(tags$h3("Cross references"), align = "center"),
              div(dropdown(
                style = "unite", size = "lg", icon = icon("crosshairs"), status = "royal", width = "300px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
                tooltip = tooltipOptions(title = "Click to see available options!"),
                h5("If you specify a Gene or a SNP id, you will find here links to GeneCards and dbSNP."),
                uiOutput(outputId = "genecards_link"),
                hr(),
                uiOutput(outputId = "gwascat_link"),
                hr(),
                uiOutput(outputId = "ld_hub_link"),
              ), align = "center"),
            ),
            div(h6("ss"), style = "color: #f2f2f2", align = "center"),
            div(h6("ss"), style = "color: #f2f2f2", align = "center"),
            hr(),
            div(h3("Top SNPs info"), align = "center"),
            div(tableOutput('table'), align = "center", style = 'font-size:100%; overflow-x: scroll; overflow-y: scroll; height:200px'),
            bsPopover("table", title='Top SNPs table', content='Most significant SNP-associations in the main plot interface. Chr: chromosome; Position: genomic position depending on reference genome choosen; ID: variant identifier (missing if variant is not in 1000Genome or too rare); -log10(P): association p-value (in -log10 scale); Study: name of the study; Alleles: reference and alternative alleles, respectively (from 1000Genome data).', placement="right",trigger="hover", options = list(container = "body")),
            hr(),
            div(h3("GWAS Catalog SNPs"), align = "center"),
            div(tableOutput('gwascat_table'), align = "center", style = 'font-size:100%; overflow-x: scroll; overflow-y: scroll; height:200px'),
            bsPopover("gwascat_table", title='GWAS Catalog table', content='Most significant associations of SNPs and genes in the plotted region as found in the GWAS Catalog (updated on 2021-12-21).', placement="right",trigger="hover", options = list(container = "body")),
            hr(),
            div(h3("Structural Variants"), align = "center"),
            div(tableOutput('sv_table'), align = "center", style = 'font-size:100%; overflow-x: scroll; overflow-y: scroll; height:200px'),
            bsPopover("sv_table", title='Structual variants table', content='Most relevant structural variations in the region of interest and from the chosen source.', placement="right",trigger="hover", options = list(container = "body")),
          )
        ),
        column(8,
          plotOutput(outputId = "plot", height = "1300px"),
        )
      ),
      hr(),
      fluidRow(           # SECOND PLOT WINDOW AND SIDEBAR OPTIONS
        column(4,
          wellPanel(
            column(6,
              div(tags$h3("GTEx settings"), align = "center"),
              div(dropdown(
                style = "unite", size = "lg", icon = icon("gear"), status = "primary", width = "450px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
                tooltip = tooltipOptions(title = "Click to see available options!"),
                radioGroupButtons(inputId = "heat_cols", label = "Colors:", choices = c("Default", "Viridis", "Plasma", "Blues", "Reds"), status = "danger", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
                multiInput(inputId = "tissues", label = "GTEx tissues", choices = c("All tissues", gtex_tissues$V1), selected = "All tissues", options = list(enable_search = TRUE, non_selected_header = "Choose between:", selected_header = "You have selected:")),
              ), align = "center")),
            column(6,
              div(tags$h3("Download"), align = "center"),
              div(dropdown(
                style = "unite", size = "lg", icon = icon("download"), status = "warning", width = "300px", animate = animateOptions(enter = animations$fading_entrances$fadeInLeftBig, exit = animations$fading_exits$fadeOutLeftBig),
                tooltip = tooltipOptions(title = "Click to see available options!"),
                div(h5("eQTLs table"), align = "center"),
                div(downloadButton("download_eQTLsTable", "Download eQTLs"), align = "center"),
                hr(),
                div(h5("sQTLs table"), align = "center"),
                div(downloadButton("download_sQTLsTable", "Download sQTLs"), align = "center"),
                hr(),
              ), align = "center")),
            div(h6("ss"), style = "color: #f2f2f2", align = "center"),
            hr(),
            div(h3("eQTL Table"), align = "center"),
            div(tableOutput('eqtls_table'), align = "center", style = 'font-size:100%; overflow-x: scroll; overflow-y: scroll; height:200px'),
            bsPopover("eqtls_table", title='eQTL Table', content='Most signiticant Expression-Quantitative-Trait-Loci in the region of interest.', placement="right", trigger="hover", options = list(container = "body")),
            hr(),
            div(h3("sQTL Table"), align = "center"),
            div(tableOutput('sqtls_table'), align = "center", style = 'font-size:100%; overflow-x: scroll; overflow-y: scroll; height:200px'),
            bsPopover("sqtls_table", title='sQTL Table', content='Most signiticant Splicing-Quantitative-Trait-Loci in the region of interest.', placement="right", trigger="hover", options = list(container = "body")),
          ),
        ),
        column(8,
          #iheatmaprOutput(outputId = "plot2", height = "700px"),
          plotOutput("plot2", height = "700px"),
        )
      ),
      hr(),
      fluidRow(           # BOTTOM ROW WITH LOGOS AND INFO
        column(8,
          h5('snpXplorer and its annotation pipeline have been developed in collaboration with Amsterdam UMC and TU Delft.'),
          h5("Any suggestion or bug report is highly appreciated. Please email n.tesi@amsterdamumc.nl or snpxplorer@gmail.com."),
        ),
        column(2, img(src='tudelft1.png', width="90%")),
        column(2, img(src='amstUMC.jpg', width="90%")),
      ),
    ),
    tabPanel("Annotation",
      # DNA image
      fluidRow(column(12, img(src='dna.png', align="center", width="100%"))),
      # Overview of the tool
      fluidRow(column(12, img(src='AnnotateMe.png', width="100%", align="center"))),
      # Tutorial image
      fluidRow(column(12, img(src='tutorial_annot.png', align="center", width="100%"))),
      hr(),
      # Input options for user within a new fluidRow
      fluidRow(
        # text area for the input variants
        column(4,
          textAreaInput(inputId = "snp_list", label = "List of variants", value = "Paste variants of interest, then choose the correct format.\n\nPlease paste 1 variant per line, no commas or colons.\n\nPaste up to 1,000 SNPs if Analysis type is 'Gene-set enrichment analysis' and up to 10,000 SNPs in case of 'SNP-gene annotation'", height = "500px", width = "70%"),
          bsPopover("snp_list", title='Annotation input', content='Please insert your SNPs of interest here: make sure the format is correct and select the adequate format in the Input type box.', placement="right",trigger="hover", options = list(container = "body"))),
        # radio buttons for input type
        column(4,
          radioGroupButtons(inputId = "snp_list_type", label = "Input type:", choices = c("rsid (rs12345)", "chr:pos (1:12345678)", "chr pos (1 12345678)"), status = "warning", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
          bsPopover("snp_list_type", title='Input type', content='For compatibility, rsid is the preferred input. If your input is not rsid, please specify the Reference version.', placement="right",trigger="hover", options = list(container = "body")),
          hr(),
          prettyRadioButtons(inputId = "analysis_type", label = "Analysis type:", choices = c("SNP-gene annotation" = "annot", "Gene-set enrichment analysis (GSEA)" = "gsea"), icon = icon("check"), bigger = TRUE, inline = TRUE, status = "info", animation = "jelly"),
          prettyCheckboxGroup(inputId = "analysis_mode", label = "Source Enrichment (if GSEA was selected):", choices = c("Default (GO:BP)" = "default", "KEGG" = "KEGG", "Reactome" = "Reactome", "Wiki Pathways" = "wiki"), inline = TRUE, selected = c("default"), status = "default", fill = TRUE),
          bsPopover("analysis_mode", title='Source Enrichment', content='Please select here the gene-set databases for the gene-set enrichment analysis.', placement="right",trigger="hover", options = list(container = "body")),
          hr(),
          radioGroupButtons(inputId = "snp_list_reference", label = "Reference Genome:", choices = c("GRCh37 (hg19)", "GRCh38 (hg38)"), status = "danger", checkIcon = list(yes = icon("ok", lib = "glyphicon"), no = icon("remove", lib = "glyphicon"))),
          hr(),
          bsPopover("snp_list_reference", title='Reference Genome', content='If your input is different from rsid, then select the correct Reference genome version.', placement="right",trigger="hover", options = list(container = "body")),
          multiInput(inputId = "gtex_type", label = "GTEx tissues", choices = c("All_tissues", "Adipose_Subcutaneous", "Adipose_Visceral_Omentum", "Adrenal_Gland", "Artery_Aorta", "Artery_Coronary",
                                                                                                   "Artery_Tibial", "Brain_Amygdala", "Brain_Anterior_cingulate_cortex_BA24", "Brain_Caudate_basal_ganglia",
                                                                                                   "Brain_Cerebellar_Hemisphere", "Brain_Cerebellum", "Brain_Cortex", "Brain_Frontal_Cortex_BA9", "Brain_Hippocampus",
                                                                                                   "Brain_Hypothalamus", "Brain_Nucleus_accumbens_basal_ganglia", "Brain_Putamen_basal_ganglia", "Brain_Spinal_cord_cervical_c-1",
                                                                                                   "Brain_Substantia_nigra", "Breast_Mammary_Tissue", "Cells_Cultured_fibroblasts", "Cells_EBV-transformed_lymphocytes",
                                                                                                   "Colon_Sigmoid", "Colon_Transverse", "Esophagus_Gastroesophageal_Junction", "Esophagus_Mucosa", "Esophagus_Muscularis",
                                                                                                   "Heart_Atrial_Appendage", "Heart_Left_Ventricle", "Kidney_Cortex", "Liver", "Lung", "Minor_Salivary_Gland", "Muscle_Skeletal",
                                                                                                   "Nerve_Tibial", "Ovary", "Pancreas", "Pituitary", "Prostate", "Skin_Not_Sun_Exposed_Suprapubic", "Skin_Sun_Exposed_Lower_leg",
                                                                                                   "Small_Intestine_Terminal_Ileum", "Spleen", "Stomach", "Testis", "Thyroid", "Uterus", "Vagina", "Whole_Blood"), selected = "Whole_Blood", options = list(enable_search = TRUE, non_selected_header = "Choose between:", selected_header = "You have selected:"), width = "500px"),
          h6("By default, only Blood is considered. Adding multiple tissues may impact your analysis as a larger number of genes may be associated with each SNP."),
          bsPopover("analysis_type", title='Analysis type', content='SNP-gene annotation: only performs SNP-to-gene mapping (max 10.000 SNPs allowed). -- Gene-set enrichment analysis: performs SNP-to-gene mapping and gene-set enrichment analysis (max 1.000 SNPs allowed)', placement = "right", trigger = "hover", options = list(container = "body"))),
        # text area for email
        column(4,
          div(textInput(inputId = "email", label = "E-mail", value = "Type your email address...", width = "400px"), align = 'center'),
          bsPopover("email", title='Your e-mail address', content='We do not do anything with it...apart from sending your annotation results. :)', placement="left",trigger="hover", options = list(container = "body")),
          # maybe also good to put a submit button -- to avoid refreshes that are annoying and slow down application
          hr(),
          #div(actionButton(inputId = "run_annotation", label = "Submit analysis!", icon = icon("paper-plane"), width = "300px", style="color: #FFFFFF; background-color: #4472C4; border-color: #1A2C4C; font-size:150%"), align = "center"),
          div(h5("You will receive a confirmation e-mail for your analysis."), align = "center"),
          div(h5("Once analysis is done (typically less than 30 minutes), you'll receive results by e-mail."), align = "center"),
          div(h5("Please check your spam folder to get your snpXplorer results."), align = "center"),
          submitButton(text = "Submit!", icon("paper-plane"), width="250px"),
        ),
      ),
      hr(),
      # final row with images of tudelft and aumc
      fluidRow(
        column(8,
          h5("snpXplorer and its Annotation pipeline have been developed in collaboration with Amsterdam UMC and TU Delft."),
          h5("Any suggestion or bug report is highly appreciated. Please email n.tesi@amsterdamumc.nl or snpxplorer@gmail.com.")),
        column(2, img(src='tudelft1.png', width="90%")),
        column(2, img(src='amstUMC.jpg', width="90%"))
      ),
    ),
    navbarMenu("More",
      tabPanel("Studies and References",
        # DNA image
        fluidRow(column(12, img(src='dna.png', align="center", width="100%"))),
        wellPanel(h2("Studies and References", align="center"),
          h5("We are including more and more summary statistics.", align="center"),
          h5("For particular requests, please contact snpxplorer@gmail.com", align="center"),
        ),
        # References
        h3("Available summary statistic within snpXplorer"),
        tableOutput(outputId = 'table_info'),
        div(style = "margin-top: 50px"),
        h3("1000 Genome Project population codes"),
        tableOutput(outputId = 'table_info_1000G'),
        div(style = "margin-top: 50px"),
        h3("Structural variants datasets"),
        tableOutput(outputId = 'table_info_SVs'),
        # Final row with images of tudelft and aumc
        br(),
        br(),
        br(),
        fluidRow(
          column(8,
            h5("snpXplorer and its annotation pipeline have been developed in collaboration with Amsterdam UMC and TU Delft."),
            h5("Any suggestion or bug report is highly appreciated. Please email n.tesi@amsterdamumc.nl or snpxplorer@gmail.com")),
          column(2, img(src='tudelft1.png', width="90%")),
          column(2, img(src='amstUMC.jpg', width="90%")),
        ),
      ),
      tabPanel("Documentation and code",
        # DNA image
        fluidRow(column(12, img(src='dna.png', align="center", width="100%"))),
        wellPanel(h2("Documentation and code", align="center"),
          h5("Documentation can be viewed in the blox below.", align="center"),
          h5("Source code is available at Github", align="center")),
        column(4, 
          wellPanel(
            h5("Source code is freely available through our Github page."),
            uiOutput("github"),
            br(),
            h5("snpXplorer can also be installed locally in your machine, but you will need to download summary statistics yourself."))),
        column(8, uiOutput("pdf_doc_view")),
        # final row with images of tudelft and aumc
        br(),
        br(),
        br(),
        fluidRow(
          column(8,
            h5("snpXplorer and AnnotateMe have been developed in collaboration with Amsterdam UMC and TU Delft."),
            h5("Any suggestion or bug report is highly appreciated. Please email n.tesi@amsterdamumc.nl.")),
          column(2, img(src='tudelft1.png', width="90%")),
          column(2, img(src='amstUMC.jpg', width="90%")),
        ),
      ),
      tabPanel("Cite us",
        # DNA image
        fluidRow(column(12, img(src='dna.png', align="center", width="100%"))),
        wellPanel(h2("Cite us", align="center")),
        column(4, 
          wellPanel(
            h4("Please do not forget to cite us if you find snpXplorer useful for your research."),
            br(),
            h4("!! snpXplorer was published in Nucleic Acid Research"),
            uiOutput(outputId = "biorxiv_link"),
            br(),
            h4("snpXplorer functional annotation was used in the following papers:"),
            h5("Niccolo’ Tesi et al., Polygenic risk score of longevity predicts longer survival across an age-continuum. The Journals of Gerontology: Series A, , glaa289, https://doi.org/10.1093/gerona/glaa289"),
            uiOutput(outputId = "longevity_ms"),
            br(),
            h5("Niccolo’ Tesi et al., The effect of Alzheimer's disease-associated genetic variants on longevity. Front Genet., https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8724252/"),
            uiOutput(outputId = "rotation_ms"))),
        column(8, uiOutput("pdf_biorxiv_view")),
        # final row with images of tudelft and aumc
        br(),
        br(),
        br(),
        fluidRow(
          column(8,
            h5("snpXplorer and AnnotateMe have been developed in collaboration with Amsterdam UMC and TU Delft."),
            h5("Any suggestion or bug report is highly appreciated. Please email n.tesi@amsterdamumc.nl.")),
          column(2, img(src='tudelft1.png', width="90%")),
          column(2, img(src='amstUMC.jpg', width="90%")),
        ),
      ),
      tabPanel("Help",
        # DNA image
        fluidRow(column(12, img(src='dna.png', align="center", width="100%"))),
        wellPanel(
          h2("Help page", align="center"),
          h4("This page provides sample files to try out both the Exploration and the Annotation part, and a quick-start video.", align = "center")),
        # add wellpanel for the sample file of the exploration section
        column(3, 
          wellPanel(
            h4("Exploration section"),
            hr(),
            h5("For the exploration section, please provide a space- or tab-separated file with header."),
            h5("Make sure there are at least chromosome number, position and p-value. snpXplorer will try to understand the columns from the file header."),
            h5("Alternatively, you can use PLINK formatted files. Look at the examples below or download a ready-to-use dataset to try out."),
            br()
          ),
        ),
        # add wellpanel for example file
        column(2,
          h5("Example file", align="center"),
          hr(),
          h6("chr pos p"),
          h6("16 89659 0.19"),
          h6("16 90318 0.48"),
          h6("16 439873 0.28"),
          h6("16 904328 0.01"),
          h6(".. .. .."),
          hr(),
          # Button to download trial and plink-based files
          downloadButton("download_help_exploration", "Download sample file")
        ),
        # add wellpanel for example file
        column(7,
          h5("Example PLINK file", align="center"),
          hr(),
          h6("#CHROM	POS	ID	REF	ALT	A1	A1_FREQ	A1_CASE_FREQ	A1_CTRL_FREQ	MACH_R2	TEST	OBS_CT	BETA	SE	Z_STAT	P"),
          h6("17	15802438	rs12449443	G	A	A	0.07	0.06	0.07	0.99	ADD	4191	0.01	0.16	0.05	0.96"),
          h6("17	29378199	rs57278847	T	C	C	0.01	0.01	0.02	0.31	ADD	4191	16.40	18.15	0.90	0.36"),
          h6("17	43286432	rs57847	T	C	C	0.02	0.11	0.42	0.31	ADD	4191	1.40	1.15	0.70	0.16"),
          h6("17	22318299	rs5724327	T	C	C	0.31	0.61	0.42	0.11	ADD	4191	6.40	58.12	0.60	0.26"),
          h6(".. .. .."),
          hr(),
          # Button to download trial and plink-based files
          downloadButton("download_help_exploration_plink", "Download sample PLINK file")
        ),
        fluidRow(),
        hr(),
        br(),
        # add wellpanel for the sample file of the annotation section
        column(3,
          wellPanel(
            h4("Annotation section"),
            hr(),
            h5("The annotation section accepts any set of SNPs (possibly more than 1)."),
            h5("Multiple format are accepted: you can choose between rs-identifier or genomic position."),
            h5("Please see examples on the right or download a trial dataset."),
            br()
          ),
        ),
        # add wellpanel for example file
        column(3,
          h5("Example file #1 ~ chr:position", align="center"),
          hr(),
          h6("16:89659"),
          h6("16:90318"),
          h6("16:439873"),
          h6("16:904328"),
          h6(".. .. .."),
          hr(),
          # Button to download trial and plink-based files
          downloadButton("download_help_annotation1", "Download sample file"),
        ),
        # add wellpanel for example file
        column(3,
          h5("Example file #1 ~ chr position", align="center"),
          hr(),
          h6("16 89659"),
          h6("16 90318"),
          h6("16 439873"),
          h6("16 904328"),
          h6(".. .. .."),
          hr(),
          # Button to download trial and plink-based files
          downloadButton("download_help_annotation2", "Download sample file")
        ),
        column(3,
          h5("Example file #3 ~ rsID", align="center"),
          hr(),
          h6("rs7412"),
          h6("rs1234"),
          h6("rs439873"),
          h6("rs904328"),
          h6(".. .. .."),
          hr(),
          # Button to download trial and plink-based files
          downloadButton("download_help_annotation3", "Download sample file")
        ),
        fluidRow(),
        hr(),
        # add wellpanel for the youtube videos
        column(12, 
          wellPanel(
            h3("Video tutorials", align = "center"),
            h4("Have a look at our quick start videos of snpXplorer.", align = "center"),
            h4("There are tutorials about exploration section and annotation sections.", align = "center"),
          ),
        ),
        column(6,
          h4("Quick Start tutorial", align="center"),
          hr(),
          uiOutput("video1"),
          hr(),
        ),
        column(6,
          h4("Load your own file tutorial", align="center"),
          hr(),
          uiOutput("video2"),
          hr(),
        ),
        column(6,
          h4("Functional annotation of your SNP list", align="center"),
          hr(),
          uiOutput("video3"),
          hr(),
        ),
        column(6,
          h4("Discover LD patterns", align="center"),
          hr(),
          uiOutput("video4"),
          hr(),
        ),
        # final row with images of tudelft and aumc
        br(),
        fluidRow(
          column(8,
            h5("snpXplorer and its annotation pipeline have been developed in collaboration with Amsterdam UMC and TU Delft."),
            h5("Any suggestion or bug report is highly appreciated. Please email n.tesi@amsterdamumc.nl or snpxplorer@gmail.com")),
          column(2, img(src='tudelft1.png', width="90%")),
          column(2, img(src='amstUMC.jpg', width="90%")),
        ),
      )
    ),
    tabPanel("Report a bug",
      # DNA image
      fluidRow(column(12, img(src='dna.png', align="center", width="100%"))),
      wellPanel(h2("Bug Report page", align="center")),
      column(12,
        h4("snpXplorer remembers the settings of your last run. In case you found a problem in the last run, you can now directly report the bug to snpXplorer."),
        br(),
        textInput(inputId = "bug_comments", label = "What did go wrong?", value = "Please add any relevant comment here...", width="350px"),
        div(style = "margin-top: 20px"),
        div(style = "border-top: 2px solid #999999"),
        div(style = "margin-top: 20px"),
        div(actionButton(inputId = "send_bug", label = "Submit bug!", icon = icon("bug"), width = "300px", style="color: #FFFFFF; background-color: #FF2600; border-color: #771301; font-size:150%"), align = "center"),
      ),
    ),
    tabPanel("What's new",
      # DNA image
      fluidRow(column(12, img(src='dna.png', align="center", width="100%"))),
      wellPanel(h2("What's new", align="center")),
      column(12,
        h3('March 2022'),
        h4('22-03-22: Completely redesigned and reprogrammed Exploration section which allows for much more investigation.'),
        hr(),
        h3('February 2022'),
        h4("17-02-22: Annotation: added sQTL analysis."),
        hr(),
        h3('December 2021'),
        h4("24-12-21: GWAS catalog release was updated to the latest available."),
        h3('November 2021'),
        h4("11-11-21: New GWAS added: Largest GWAS of Alzheimer's disease was added."),
        h4("11-11-21: New GWAS added: Multivariate analysis of Longevity, Healthspan and Lifespan was added."),
        h4("10-11-21: New Annotation analysis added: possibility to do only SNP-gene mapping without gene-set enrichment analysis. This allows the user to annotate up to 10,000 SNPs."),
        hr(),
        h3('August 2021'),
        h4("25-08-21: CADD v1.6 (the most updated) is now used for SNP annotation."),
        h4("25-08-21: To cope with rare SNPs, we now use data from all individuals of the 1000Genome Project (before, only European individuals and common SNPs were recognized)."),
        hr(),
        h3('July 2021'),
        h4("25-07-21: New GWAS added: GR@ACE GWAS of Alzheimer's disease was added."),
        hr(),
        h3('May 2021'),
        h4('19-05-21: snpXplorer was born!'),
      ),
    )
  )
